### PR TITLE
[ModelRunnerV2] Avoid pipeline parallel bubbles

### PR DESCRIPTION
**Pull request by njhill**
Reorganize PP scheduling and broadcast of sampled tokens so that pipelining works properly for decode and (chunked) prefill.

Decode tokens for a given request are now scheduled every `pp_size` steps, the broadcast of sampled tokens from last stage to prior stages happens using a different stream and process group, so it can happen in parallel with consecutive stage p2p comms. For async scheduling we schedule `pp_size + 1` steps at a time.

Prefill chunks are pipelined in consecutive steps.

[This animation](https://github.com/user-attachments/files/28326807/async_pp_animation.html) shows the data flow between ranks for hidden states and broadcasted sample tokens for decode requests.

image

#### Benchmarks

  Model: MiniMax-M2.7 (FP8)
  Hardware: Single GB200 node, PP=4, TP=1, DP=1
  Server: vllm serve with `--compilation-config '{"mode":3,"pass_config":{"fuse_minimax_qk_norm":true}}'`, `--no-enable-prefix-caching`
  Client: vllm-bench, random dataset, `max_concurrency=128`, `--ignore-eos`
  Per config: 1 warmup batch + 3 timed runs (mean reported)

  | Workload (isl/osl) | total tok/s gain | TTFT reduction |
  | ----------------- | ---------------- | --------- |
  | 1024 / 1024       | 1.24x            | 2.34x     |
  | 128000 / 1        | 3.17x            | 3.15x     |
  | 80000 / 1024      | 2.28x            | 3.11x     |

It also seems that this fixes an IMA that was previously occurring often with the mixed workload (1k/1k).

### CODE REVIEW

File: vllm/v1/core/sched/async_scheduler.py [hunk 1]
```diff
@@ -34,6 +36,9 @@ def _update_after_schedule(self, scheduler_output: SchedulerOutput) -> None:
             # We will update the actual spec token ids in the worker process.
             request.spec_token_ids = self._spec_token_placeholders
 
+            if self.pp_throttle:
```
Comment (WoosukKwon) [inline vllm/v1/core/sched/async_scheduler.py]: technically this is generalizable to pp_size=1, right?

File: vllm/v1/worker/gpu/model_states/interface.py [hunk 2]
```diff
@@ -57,9 +57,7 @@ def apply_staged_writes(self) -> None:
         return None
 
     def postprocess_state(
-        self,
-        input_batch: InputBatch,
-        num_sampled: torch.Tensor,
+        self, idx_mapping: torch.Tensor, num_sampled: torch.Tensor
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_states/interface.py:60]: Can you please explain about this change? I thought passing `input_batch` as a whole is better in terms of API stability

File: vllm/v1/worker/gpu/states.py [hunk 3]
```diff
@@ -87,12 +90,14 @@ def add_request(
         prompt_len: int,
         all_token_ids: list[int],
         num_computed_tokens: int,
+        max_tokens: int = 0,
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/states.py]: nit: I think we shouldn't provide any default here

File: vllm/v1/worker/gpu/model_runner.py [hunk 4]
```diff
@@ -701,11 +722,13 @@ def add_requests(self, scheduler_output: SchedulerOutput) -> None:
             self._remove_request(req_id)
 
             prompt_len = len(new_req_data.prompt_token_ids)
+            sampling_params = new_req_data.sampling_params
             self.req_states.add_request(
                 req_id=req_id,
                 prompt_len=prompt_len,
                 all_token_ids=new_req_data.prefill_token_ids,
                 num_computed_tokens=new_req_data.num_computed_tokens,
+                max_tokens=sampling_params.max_tokens if sampling_params else 0,  # type: ignore[arg-type]
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Should this be 1 instead of 0?

File: vllm/v1/worker/gpu/model_runner.py [hunk 5]
```diff
@@ -978,12 +1001,13 @@ def sample(
...
 
-    def postprocess(
+    def postprocess_sampled(
         self,
-        input_batch: InputBatch,
+        idx_mapping: torch.Tensor,
         sampled_tokens: torch.Tensor,
         num_sampled: torch.Tensor,
         num_rejected: torch.Tensor,
+        query_start_loc: torch.Tensor | None = None,
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py:1051]: When is this `None`?

File: vllm/v1/worker/gpu/model_runner.py [hunk 6]
```diff
@@ -1246,20 +1276,34 @@ def sample_tokens(
...
+            # Optimistically update num_computed_tokens for entire batch here.
+            # Will be adjusted for rejections if necessary in update_requests.
+            self.postprocess_num_computed_tokens(input_batch)
+            if not all_decode_next:
+                # Might contain non-final prefill chunks, which will be scheduled
+                # in the immediate next step (rather than in pp_size steps).
+                # TODO(nick) this is just for mamba, see if we can simplify
+                self.model_state.postprocess_state(
+                    input_batch.idx_mapping, torch.zeros_like(input_batch.idx_mapping)
+                )
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Dumb question: Why does Mamba need the special handling?

File: vllm/v1/worker/gpu/model_runner.py [hunk 7]
```diff
@@ -1157,12 +1182,17 @@ def execute_model(
             assert intermediate_tensors is not None
             assert self.intermediate_tensors is not None
             n = input_batch.num_tokens_after_padding
-            model_inputs["intermediate_tensors"] = IntermediateTensors(
-                {
+            if dummy_run:
+                # PP + DP case.
+                tensors = {
+                    k: v[:n] for k, v in self.intermediate_tensors.tensors.items()
+                }
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Why do we need this separate branch?

File: vllm/v1/worker/gpu/pp_utils.py [hunk 8]
```diff
@@ -2,40 +2,220 @@
...
-    torch.distributed.broadcast(sampled_tokens, src=pp.last_rank, group=pp.device_group)
+            to_record = (sampled_token_ids, num_sampled, num_rejected)
+            if not need_sampled_mask.all():
+                subset_indices_np = np.flatnonzero(need_sampled_mask).astype(np.int32)
+                subset_indices = torch.from_numpy(subset_indices_np).to(
+                    self.device, non_blocking=True
+                )
+                sampled_token_ids = sampled_token_ids[subset_indices]
+                num_sampled = num_sampled[subset_indices]
+                num_rejected = num_rejected[subset_indices]
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/pp_utils.py]: Why not send all?

File: vllm/v1/worker/gpu/model_runner.py (hunk 7 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_runner.py]: In the [dummy run case](https://github.com/vllm-project/vllm/pull/42187/changes#diff-5823f988fc0264681a80db24ccaba4d364f14394815d83ec1d90944c09f571f0R505-R507), we pass in the persistent `self.intermediate_tensors` as the input tensors so the copy in the normal branch fails (src and dest are same memory).

File: vllm/v1/worker/gpu/model_states/interface.py (hunk 2 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_states/interface.py:60]: In the PP case we defer postprocessing requests whose next scheduled step will be decode (excludes non-final prefill chunks and requests that have reached `max_length`), which will be a subset of the batch.

In those cases, this method is called in a later step at which point the input batch may have different contents.

So we just store the [filtered](https://github.com/vllm-project/vllm/pull/42187/changes#diff-8330021c29feb32718335218ea994b6a770dd182a94b1baeacb7f7f42214afb7R139-R146) `idx_mapping` which corresponds to the subset of reqs whose state needs updating with sampled tensors.

File: vllm/v1/worker/gpu/model_runner.py (hunk 4 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_runner.py]: The 0 means more like N/A for pooling models, functionally it won't make a difference here, I could change it to 1 also.

File: vllm/v1/worker/gpu/model_runner.py (hunk 6 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_runner.py]: The mamba integration has a `postprocess_state` hook which uses `num_sampled`: https://github.com/vllm-project/vllm/blob/e54eff769dea26f00b122be89ebaa870875da03e/vllm/v1/worker/gpu/model_states/mamba_hybrid.py#L141

File: vllm/v1/worker/gpu/pp_utils.py (hunk 8 above)
Comment (njhill) [reply vllm/v1/worker/gpu/pp_utils.py]: We need to do the same filtering/re-allocation on the receive side otherwise, so thought better to do here and send smaller tensors.

The filtering is needed so that the `post_update` kernel will only update the "deferred" subset of requests

File: vllm/v1/worker/gpu/model_runner.py (hunk 5 above)
Comment (njhill) [reply vllm/v1/worker/gpu/model_runner.py:1051]: I'm now incrementing the computed tokens by the query lens unconditionally for the whole batch in the immediate postprocessing ([here](https://github.com/vllm-project/vllm/pull/42187/changes#diff-5823f988fc0264681a80db24ccaba4d364f14394815d83ec1d90944c09f571f0R1281-R1283)).

Which means in the deferred postprocessing (once we know the accepted and rejected counts), we just need to subtract the rejected count.

So the query_start_locs aren't needed and omitted when this is called in the deferred case, and the `post_update` kernel is [modified](https://github.com/vllm-project/vllm/pull/42187/changes#diff-106f39c08266f186830bb8fcd7fb1df35c0aaa5fd0ac5c17aac64aeddee48902R467-R478) to only subtract the rejected count when `query_start_loc` is passed as `None`.

**Comment by njhill** [comment]
@WoosukKwon thanks for the review so far. Looking through the changes again now, I realize it's still fairly complicated. I can spend some more time trying to simplify further.

File: vllm/v1/worker/gpu/pp_utils.py (hunk 8 above)
Comment (njhill) [reply vllm/v1/worker/gpu/pp_utils.py]: I guess it could end up simpler to just slice on the receive side, I'll try that

**Comment by njhill** [comment]
This is now ready but depends on https://github.com/vllm-project/vllm/pull/44277 or similar.

File: vllm/v1/worker/gpu/model_runner.py [hunk 9]
```diff
@@ -1155,12 +1191,13 @@ def execute_model(
...
-                    k: v[:n].copy_(intermediate_tensors.tensors[k][:n])
-                    for k, v in self.intermediate_tensors.tensors.items()
-                }
-            )
+            new_tensors = {
+                k: v[:n]
+                if dummy_run
+                else v[:n].copy_(intermediate_tensors.tensors[k][:n])
+                for k, v in self.intermediate_tensors.tensors.items()
+            }
```
Comment (njhill) [inline vllm/v1/worker/gpu/model_runner.py:1229]: I changed this to a single comprehension since it's more concise, but for performance better to have a separate one for `dummy_run`. Let me know what is preferred.

File: vllm/v1/worker/gpu/model_runner.py [hunk 10]
```diff
@@ -999,12 +1042,13 @@ def sample(
         )
         return sampler_output, num_sampled, num_rejected
 
-    def postprocess(
+    def postprocess_sampled(
         self,
-        input_batch: InputBatch,
+        idx_mapping: torch.Tensor,
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu/model_runner.py]: Maybe it's good to comment here too?

```suggestion
        idx_mapping: torch.Tensor,  # May include -1 for masked entries
```

**Review by WoosukKwon — APPROVED** [review]
LGTM. Thanks for the fix!
